### DISCUSSION TITLE
Is it possible to build android apk with buildozer or similar?

### DISCUSSION THREAD

**Discussion by bobwatcherx** [zauberzeug/nicegui#578] [Ideas / Feature Requests]
What if I want nicegui to run for android . And is it possible to use buildozer way to build android apk . And are there any future plans for Nicegui to support mobile as well. Thanks

**Comment by rodja** [comment]
Yes, we would love to be able to run NiceGUI as packaged Apps on Android and iOS. We have not yet looked into it but https://buildozer.readthedocs.io/ may be a good candidate. It would be great if someone could try it.

**Comment by cj** [comment]
I've been looking into this too. It might be worth doing something similar to what the Jumpstart team did, https://jumpstartrails.com/ios. They created a native wrapper for both IOS and Android which just uses a web view but gives you access to certain native functionality.

**Comment by igrowing** [comment]
Another option to implement build for Android is [Briefcase](https://briefcase.readthedocs.io/en/latest/how-to/publishing/android.html). It does not require explicit usage of Android studio, all the process goes stealth. However, not clear how to tackle the iOS builds then.

**Reply by ItsCubeTime** [reply]
Beeware supposedly has iOS support as well

**Reply by ItsCubeTime** [reply]
Ive been doing some experiments trying to get nicegui to run in a toga app on android, however Ive been unsuccessful so far. Toga + nicegui runs great on my Windows desktop device however (as a pywebview alternative)

**Comment by traverseda** [comment]
PyWebView explicitly support buildozer, so there's no reason I can see this wouldn't work with stock buildozer config. Not seeing any weird C API stuff as hard dependencies other than gunicorn, which is pretty common and probably not a *hard* dependency.

**Comment by jeff-dh** [comment]
I just tried to get nicegui running on android with buildozer with the following results:

-  for some reason (unknown to me atm) the nicegui dependencies need to be explicitly added to buildozer.spec file, otherwise the android app throws import errors:
    `requirements = python3,nicegui,typing-extensions,python-socketio,asyncio,python-engineio,bidict,starlette,pydantic,anyio,sniffio,itsdangerous,aiofiles,aiohttp,vbuild`
- afterwards I get:
    `NotImplementedError: This Python build lacks multiprocessing.synchronize, usually due to named semaphores being unavailable on this platform.`
- seams to make no difference whether bootstrap is set to webview or not, whether `reload` is `True` or `False`, nor whether `native` is `True` or `False`

Anyone any ideas about the `multiprocessing.synchronize NotImplementedError`?

Did anyone get any nicegui app running on android yet?

**Reply by jeff-dh** [reply]
https://docs.python.org/3/library/intro.html#mobile-availability

> - On Android, creating subprocesses is possible but [officially unsupported](https://issuetracker.google.com/issues/128554619#comment4). In particular, Android does not support any part of the System V IPC API, so [multiprocessing](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing) is not available.

**Reply by evnchn** [reply]
Not sure if FastAPI / uvicorn under the hood require it, though, so perhaps you can first do with a minimal FastAPI app / minimal uvicorn app. 

If so, then the question to ask is, can we possibly rip out any part of NiceGUI which creates subprocesses?

**Reply by jeff-dh** [reply]
```py
from fastapi import FastAPI
import uvicorn

app = FastAPI()


@app.get("/")
async def root():
    return {"message": "Hello World"}

uvicorn.run(app, port=5000, log_level="info")
```

That works...... app starts, shows blank screen, no errors and  can open localhost:5000 and get the response

**Reply by evnchn** [reply]
Then it's on us. Will look into it once I am not so tired.

**Reply by jeff-dh** [reply]
btw: `requirements = python3,fastapi,uvicorn,typing-extensions,python-socketio,asyncio,python-engineio,bidict,starlette,pydantic,anyio,sniffio,itsdangerous,aiofiles,aiohttp,vbuild,click,h11` for minimal fastapi / uvicorn app

**Reply by rodja** [reply]
Whats your minimal NiceGUI app? Make sure you start with `reload=False` otherwise NiceGUI starts in a subprocess.

**Reply by evnchn** [reply]
He mentioned: 
> seams to make no difference whether bootstrap is set to webview or not, **whether reload is True or False**, nor whether native is True or False

**Reply by rodja** [reply]
There is a difference. If either `native` or `reload` is set to `True` NiceGUI uses multiprocessing. If that is not supported on Android all further experiments should take that into account.

**Reply by jeff-dh** [reply]
```python
from nicegui import ui

ui.label('Hello World')

ui.run(port=8000, reload=False)
```
<img src="https://github.com/user-attachments/assets/8469c4e3-933a-4774-8754-3165a30ff458" width=600/>



it seems to be about named semaphores, not sure whether sub processes really are an issue (searching the web suggests that sub processes might work).

**Reply by jeff-dh** [reply]
explicitly setting `native=False` makes no difference

**Comment by jeff-dh** [comment]
Got it working on my phone!

The nicegui issues are the ThreadPoolExecutor and ProcessPoolExecutor for io and cpu bound tasks in run.py. The rest is buildozer config.

I'll give you an detailed heads up later or tomorrow.

**Comment by jeff-dh** [comment]
https://github.com/jeff-dh/nicegui/tree/android (fc4e5818b88f7cf4cc5fd72e94c92e6b81300d2a)
these are probably the minimal necessary changes to make it run on android. It might make sense to add a couple more (i.e. welcome.py crashes when `host!=0.0.0.0` due to the `io_bound` call).

It's a little hack-ish to use the nicegui fork (and not the pypi package) in the buildozer build. It works with a symbolic link in the source dir to the patched nicegui package (which overrides the required pypi package).
-> niceApp/README.md

an example app (with `buildozer.spec` file):
https://github.com/jeff-dh/niceApp

```python
from nicegui import ui, app

def is_android():
    import platform
    return "android" in platform.uname().release

@ui.page("/")
def index():
    if is_android():
        ui.label("Hello nice android world")

    storage = app.storage.user
    storage["foobar"] = storage.get("foobar", 0)

    def callback():
        storage["foobar"] += 1
        ui.label(f"yeah{'.'*storage['foobar']}")

    ui.button("Really?", on_click=callback)

ui.run(host='localhost', reload=False, show=False, storage_secret="foobar")
```

works as expected as a webview app on my phone.

If you have any questions, let me know.

**Reply by evnchn** [reply]
@jeff-dh Looks goot to merge except for three issues: 

1. ~~Drop `assert process_pool is not None` from line 87 in `def tear_down()`, otherwise it will error out.~~ Check my comment below for a full solution. 
2. Add back `  # pylint: disable=global-statement # noqa: PLW0603` for `global process_pool` to make the code check stop complaining. 
3. `if not process_pool:` is not preferred (in case `process_pool` is falsy?). Use `if process_pool is None:` is more robust and clear. Also for `thread_pool`

**Reply by evnchn** [reply]
Should be something like: 

```py
def tear_down() -> None:
    """Kill all processes and threads."""
    if helpers.is_pytest():
        return
    kwargs = {'cancel_futures': True} if sys.version_info >= (3, 9) else {}
    if process_pool is not None:
        for p in process_pool._processes.values():  # pylint: disable=protected-access
            p.kill()
        process_pool.shutdown(wait=True, **kwargs)
    if thread_pool is not None:
        thread_pool.shutdown(wait=False, **kwargs)
```

---

Does not mean it is totally ready to merge. You'll have to PR to find out what others think.

**Reply by jeff-dh** [reply]
9ae98d9c91c84aabe472512e74c82c447d1af73d

**Reply by jeff-dh** [reply]
I'm wondering whether it makes sense to make `io_bound` and `cpu_bound` work on android.

This could probably be achieved by either 
- somehow run the task in the same thread (e.g. disable the underlying feature `if platform == android`)
- or by trying to use some kind of android compatible concurrent execution instead of `multiprocessing.*ExecutorPool`. 

This would make `welcome.py` work if `host!="0.0.0.0"` and -- I assume -- would actually be pretty handy for android apps....

**Reply by evnchn** [reply]
@jeff-dh go ahead and do a PR. 

> This PR fix #578 and enables nicegui work on android with buildozer, provided that they do not use `run.io_bound` or `run.cpu_bound` in any capacity. Thus, buildozer compatibility is brought in line with FastAPI, which works on android with buildozer without modifications.

---

About `host must not equal "0.0.0.0"`: 

Culprit: 

https://github.com/zauberzeug/nicegui/blob/2b8b1a696992e64578f2e4c5fc285e90b60b87a6/nicegui/welcome.py#L22

My idea is: 

```py
    try:
        ips = set((await run.io_bound(_get_all_ips)) if host == '0.0.0.0' else [])
    except Exception as e:
        ips = set(_get_all_ips() if host == '0.0.0.0' else [])
```

Brutal, but it works. 

---

No comment on `reload=True`. Who would do that? Is it too much hand-holding, if we coerce `reload=False` if detect on Android?

**Reply by evnchn** [reply]
@jeff-dh I think we should separate the two issues. 

The first one is a necessity one: NiceGUI flat-out won't run

The second one is a nice-to-have one: getting `run.io_bound` and `run.cpu_bound` to work, after NiceGUI runs. 

IMO There is no need to hold up the first issue, because we need more time to figure out the second one. 

Sidenote: really want to get this to work, so this absolutely piques my interest.

**Reply by jeff-dh** [reply]
> My idea is:
> 
> ```python
>     try:
>         ips = set((await run.io_bound(_get_all_ips)) if host == '0.0.0.0' else [])
>     except Exception as e:
>         ips = set(_get_all_ips() if host == '0.0.0.0' else [])
> ```
> 
> Brutal, but it works.

I don't really like this, because this adds a special case for an issue we're about to tackle.... therefor I would prefer do a pr for the work so far. This would  add basic android compatibility upstream (under certain conditions, `host != "0.0.0.0"`) and hopefully harms no one.
I'm fine with handling `io_bound` & `cpu_bound` for android in another step.

**Reply by evnchn** [reply]
Sure. Perhaps we can let the preliminary Android compatibility be upstream at 2.16.0 in the upcoming release. 

It'll also make devwork on Android compatibility easier, lowering the barrier to entry. 

Now I suggest you to do the PR as well as enable allow edit by maintainer to speed things along. 

And I agree it harms no one. I don't see how the new code would diverge in behavior to the old code.

**Reply by jeff-dh** [reply]
#4683 
(hmmmm auto linking not working for some reason, sorry)

**Comment by jeff-dh** [comment]
```python
# run_android.py
from .run import P, R, Callable

import asyncio
from multiprocessing import Pipe, Process

async def cpu_bound(callback: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
    """Run a CPU-bound function in a separate process....."""

    def wrapper(pipe, f, args, kwargs):
        pipe.send(f(*args, **kwargs))

    parent_pipe, child_pipe = Pipe()

    p = Process(target=wrapper, args=(child_pipe, callback, args, kwargs))
    p.start()

    return await asyncio.to_thread(parent_pipe.recv)


async def io_bound(callback: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
    """Run an I/O-bound function in a separate thread."""

    return await asyncio.to_thread(callback, *args, **kwargs)
```
https://docs.python.org/3/library/asyncio-task.html#running-in-threads

seems to work on android

**Reply by evnchn** [reply]
1. Can we really use asyncio.to_thread() for CPU-bound tasks? Since there is still the GIL. It doesn't seem possible to run CPU-bound tasks without blocking up the rest of the async code, without putting said task in another process, which I'm not sure if your implementation does appropriately. 

> Note: Due to the GIL, asyncio.to_thread() can typically only be used to make IO-bound functions non-blocking. However, for extension modules that release the GIL or alternative Python implementations that don’t have one, asyncio.to_thread() can also be used for CPU-bound functions.

A demonstration program which runs a CPU-bound task for 10 second, while we observe whether the NiceGUI app freezes up, can be a way to answer. 

2. I don't think we should override the old implementation, thus affecting a sizeable chunk of existing users. Due to how it already consumes args and kwargs, maybe we can have `run.asyncio_io_bound` and `run.asyncio_cpu_bound` for the new implementation?

**Reply by evnchn** [reply]
Or rather, `run_asyncio.cpu_bound` and `run_asyncio.io_bound`

Also, the existing run functions can call the new ones call if error.

**Reply by jeff-dh** [reply]
1. if I'm not wrong: `cpu_bound` does not use `asyncio.to_thread` for a cpu bound call, it uses `to_thread` to await `pipe.recv` which is an IO bound function, right? Anyway, I'm not an asyncio expert and I'm not sure whether that's a feasible approach.
2.  what about something like this or similar:
```python
#run.py
import plaform

if not 'android' in platform.uname().release:
    from run_regular import io_bound, cpu_bound
else:
    from run_android import io_bound, cpu_bound
```
3. TODO or consider: What about error handling? Does it satisfy all requirements? Is there any necessity to use pools? ....?

**Reply by evnchn** [reply]
1. Ah! Then I understand it. In principle it should work, but I am still worried about janky stuff, such as the Android implementation of `Process` and `Pipe` not actually opening a separate process, but some kind of software trickery to make those calls work despite Android's limitations. The "run a 10-second CPU bound task while checking if NiceGUI stays responsive" still seems the most "solid and empirical" solution to me
2. Again, would not recommend checking explicitly platform name android. You know that `Buildozer supports packaging for iOS via Kivy iOS`, right? 
3. This is why I would more inclined to providing `run_asyncio` for Android and the likes where the normal `run` does not work, while we pass calls from the old `run` to the new `run_asyncio` if it fails **with a warning that users should explicitly change to the new one if they want to support both Desktop, and platforms which does not support `ProcessPoolExecutor` and/or `ThreadPoolExecutor` such as Android**. By getting the user to switch to the new `run_asyncio` if they want their code to work on both Android and Desktop, we don't need to worry about aligning the two functions exactly in their inner workings, as the user is expected to check for minor inconsistencies on the desktop side when making the switch, not us grinding `run_asyncio` to walk like `run` and quack like `run`.

**Reply by jeff-dh** [reply]
Or:

- we could use the setup function approach you suggested in #4683
- and do something like this:
```python
async def cpu_bound(callback: Callable[P, R], *args: P.args, **kwargs: P.kwargs) -> R:
    if not process_pool is None:
        return await _run(process_pool, safe_callback, callback, *args, **kwargs)
    else:
        print("WARNING!")
        return run_asyncio.cpu_bound(callback, *args, **kwargs)
```

**Reply by evnchn** [reply]
I think the setup function approach that I suggested is necessary to be implemented for that PR to be 100% backwards compatible in terms of behaviour and speed. So I think you can include it before the review, saving one round of back and forth. 

With that, this `cpu_bound` implementation makes total sense. Warning you should use `log.warning`. Look it up in the codebase. 

I am not sure if this should go in the same PR or a separate one, though. We can make a separate one, because this new cpu_bound would behave 100% like the old one as `process_pool` is always not None (same goes for `io_bound`), no need to be afraid of being blocked by the other PR. 

Also, I think the preferred order is `if process_pool is not None:`. Check the codebase for many instances of `is not None:`

**Comment by jeff-dh** [comment]
with either of these modules

- [`run_simple`](https://github.com/jeff-dh/nicegui/blob/android_cpu_bound/nicegui/run_simple.py)
a simple proof of concept android compatible cpu_bound implementation

- [`run_process_pool`](https://github.com/jeff-dh/nicegui/blob/android_cpu_bound/nicegui/run_process_pool.py)
an android compatible cpu_bound implementation using a custom process pool based on the concept from `run_simple`

this works as expected

```python
from nicegui import ui
# from nicegui.run_simple import cpu_bound
from nicegui.run_process_pool import cpu_bound, shutdown

def fib(n):
    assert n >= 0
    if n == 0: return 0
    if n <= 2: return 1
    return fib(n-1) + fib(n-2)


@ui.page("/")
async def index():
    async def fib_callback(x):
        try:
            lbl.text = await cpu_bound(fib, x.value)
        except Exception as e:
            lbl.text = e.__repr__()

    ui.number(label="fib", on_change=fib_callback, value=0)
    lbl = ui.label("--")
    ui.button("Really?", on_click=lambda: ui.label(f"yeah"))

try:
    ui.run(reload=False, show=False)
finally:
    shutdown()
```
tested on Linux & Android, exception handling from sub processes also works

**Reply by evnchn** [reply]
That's great!

Still, you may want to use `SubprocessException` and `safe_callback` to ensure exceptions are picklable, just like how the existing `run.cpu_bound` and `run.io_bound` does.

**Reply by jeff-dh** [reply]
https://github.com/jeff-dh/nicegui/blob/533fa7b2a6e5450bfd900fb094fff347abd555c4/nicegui/run_process_pool.py#L19-L23

**Reply by evnchn** [reply]
Aha! I was searching for `safe_callback` and didn't saw the `SubprocessException`.

**Comment by evnchn** [comment]
In short: the answer will be yes because of #4683, and available when 2.17.0 comes around. 

With the exception that the `run.cpu_bound` is not available, potential patches are at https://github.com/zauberzeug/nicegui/discussions/578#discussioncomment-13031519
